### PR TITLE
Initialized  Google Auth in backend

### PR DESCRIPTION
**Pull request by manav1403**
**End Point**:: /defaults/googleauth

### CODE REVIEW

File: controllers/default.py [hunk 1]
```diff
@@ -976,6 +976,38 @@ def user():
...
+            raise HTTP(401, "AUTH rejected")
+        token = resp.json()["id_token"]
+        return redirect(utilities.gauth_redirect(token))
+    else:
+        auth_uri = ("https://accounts.google.com/o/oauth2/v2/auth?response_type=code"
+                    "&client_id={}&redirect_uri={}&scope={}").format(CLIENT_ID, REDIRECT_URI, SCOPE)
+        return redirect(auth_uri)
+
+# ----------------------------------------------------------------------------
+
```
Comment (raj454raj) [inline controllers/default.py:1010]:
```suggestion
```

File: controllers/user.py [hunk 2]
```diff
@@ -40,6 +40,69 @@ def login_token():
...
+    """
+    auth_token = request.vars.get('g_token')
+    email = request.vars.get('email')
+
+    #verify parameters
+    if not email or not auth_token:
+        raise HTTP(400, 'Invalid parameters')
+        return
+    #verify credentials
+    user_info = json.loads(current.REDIS_CLIENT.get("g_token_"+email))
```
Comment (raj454raj) [inline controllers/user.py:56]: what if the token is not found? (Please handle it)

File: controllers/user.py [hunk 3]
```diff
@@ -40,6 +40,69 @@ def login_token():
...
+    atable.email.default = email
+    atable.first_name.default = user_info['first_name']
+    atable.last_name.default = user_info['last_name']
+
+    form = SQLFORM(db.auth_user,
+                   fields=form_fields,
+                   showid=False)
+
+    if form.process(onvalidation=current.sanitize_fields).accepted:
+        current.REDIS_CLIENT.delete("g_token_"+email)
```
Comment (raj454raj) [inline controllers/user.py:93]:
```suggestion
        current.REDIS_CLIENT.delete("g_token_" + email)
```

File: controllers/user.py [hunk 4]
```diff
@@ -40,6 +40,69 @@ def login_token():
...
+        print('pp')
+        user = db.auth_user(**{"email": email})
+        print(user)
+        auth.login_user(user)
+        return redirect('/dashboard')
```
Comment (raj454raj) [inline controllers/user.py:98]: redirect(URL(...))

File: modules/utilities.py [hunk 5]
```diff
@@ -56,6 +56,7 @@ def verifier(*args, **kwargs):
             if token not in allowed_tokens:
                 current.response.status = 401
                 return current.response
+            current.request.extension = 'json'
```
Comment (raj454raj) [inline modules/utilities.py:59]: why is this required?

File: modules/utilities.py [hunk 6]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
     return verifier
 
 # -----------------------------------------------------------------------------
+def gauth_redirect(token):
+    import requests
+    from gluon import URL
```
Comment (raj454raj) [inline modules/utilities.py:76]: isn't this already imported?

File: modules/utilities.py [hunk 7]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
...
+    user = current.db.auth_user(**{"email": user_email})
+    if not user :
+        user_info = {
+            'g_token':data['jti'],
+            'first_name': data['given_name'],
+            'last_name': data['family_name']
+        }
+        user_info = json.dumps(user_info)
+        current.REDIS_CLIENT.set('g_token_'+user_email, user_info, ex=1 * 60 * 10)
+        return '/user/fill_details?g_token='+data['jti']+'&email='+user_email
```
Comment (raj454raj) [inline modules/utilities.py:89]: use URL helper to generate this

File: modules/utilities.py [hunk 8]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
     return verifier
 
 # -----------------------------------------------------------------------------
+def gauth_redirect(token):
+    import requests
+    from gluon import URL
+    resp = requests.post("https://oauth2.googleapis.com/tokeninfo", data = {'id_token':token})
```
Comment (raj454raj) [inline modules/utilities.py:77]:
```suggestion
    resp = requests.post("https://oauth2.googleapis.com/tokeninfo", data={"id_token": token})
```

File: modules/utilities.py [hunk 9]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
...
+    data = resp.json()
+    user_email = data['email']
```
Comment (raj454raj) [inline modules/utilities.py:79]: use double quotes in python

File: modules/utilities.py [hunk 10]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
...
+    user = current.db.auth_user(**{"email": user_email})
+    if not user :
+        user_info = {
+            'g_token':data['jti'],
```
Comment (raj454raj) [inline modules/utilities.py:83]:
```suggestion
            'g_token': data['jti'],
```

File: modules/utilities.py [hunk 11]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
...
+            'first_name': data['given_name'],
+            'last_name': data['family_name']
+        }
+        user_info = json.dumps(user_info)
+        current.REDIS_CLIENT.set('g_token_'+user_email, user_info, ex=1 * 60 * 10)
+        return '/user/fill_details?g_token='+data['jti']+'&email='+user_email
+    current.auth.login_user(user)
+    return URL("dashboard")
```
Comment (raj454raj) [inline modules/utilities.py:91]:
```suggestion
    return URL("default", "dashboard")
```

File: controllers/user.py [hunk 12]
```diff
@@ -40,6 +40,69 @@ def login_token():
...
+
+    for site in current.SITES:
+        form_fields.append(site.lower() + "_handle")
+
+    atable = db.auth_user
+    stable = db.submission
+
+    # Set defaults
+    atable.stopstalk_handle.comment = T("StopStalk handle cannot be empty")
+    atable.stopstalk_handle.requires.append(IS_NOT_EMPTY(
```
Comment (raj454raj) [inline controllers/user.py:77]: can we put this validation in db.py while defining the model itself?

File: views/default/user.html [hunk 13]
```diff
@@ -19,6 +24,10 @@ <h2>
         {{pass}}
     </div>
     <div id="web2py_user_form" class="col s8">
+    <a href="/default/googleauth">
+        <img src="{{=URL('static','images/btn_google_sigin.png')}}" height="50px">
```
Comment (raj454raj) [inline views/default/user.html:28]: Can we use the Google documentation's html instead of the image?
They have a bunch of guidelines to adhere to which is better if they update on their own with minimal effort from our side.

Comment (raj454raj) [reply views/default/user.html:28]: https://developers.google.com/identity/sign-in/web/build-button

**Review by raj454raj — CHANGES_REQUESTED** [review]
Yet to test this locally, have added basic things on the PR.

File: controllers/user.py (hunk 2 above)
Comment (manav1403) [reply controllers/user.py:56]: if the token is not found it will be null
and according to if the condition on line 57 it will not be equal to not nullable field auth_token and thus an error will be thrown back

File: views/default/user.html (hunk 13 above)
Comment (manav1403) [reply views/default/user.html:28]: the button not at all looks good and the button used is downloaded from google documentation only

**Comment by manav1403** [comment]
@raj454raj pls review

File: controllers/user.py [hunk 14]
```diff
@@ -40,6 +40,66 @@ def login_token():
...
+                   "stopstalk_handle"]
+
+    for site in current.SITES:
+        form_fields.append(site.lower() + "_handle")
+
+    atable = db.auth_user
+    stable = db.submission
+
+    # Set defaults
+    atable.stopstalk_handle.comment = T("StopStalk handle cannot be empty")
```
Comment (raj454raj) [inline controllers/user.py:76]: Is this required?

File: controllers/user.py [hunk 15]
```diff
@@ -40,6 +40,66 @@ def login_token():
...
+
+    atable.email.readable = True
+    atable.email.writable = False
+    atable.email.comment = T("Email cannot be updated")
```
Comment (raj454raj) [inline controllers/user.py:80]: Is this required?

File: controllers/user.py [hunk 16]
```diff
@@ -40,6 +40,66 @@ def login_token():
...
+    atable.last_name.default = user_info['last_name']
+
+    form = SQLFORM(db.auth_user,
+                   fields=form_fields,
+                   showid=False)
+
+    if form.process(onvalidation=current.sanitize_fields).accepted:
+        current.REDIS_CLIENT.delete("g_token_" + email)
+        user = db.auth_user(**{"email": email})
+        print(user)
```
Comment (raj454raj) [inline controllers/user.py:93]: delete

File: modules/utilities.py [hunk 17]
```diff
@@ -70,6 +71,25 @@ def verifier(*args, **kwargs):
...
+    user = current.db.auth_user(**{"email": user_email})
+    if not user:
+        user_info = {
+            'g_token': data['sub'],
+            'first_name': data['given_name'],
+            'last_name': data['family_name']
+        }
+        user_info = json.dumps(user_info)
+        current.REDIS_CLIENT.set('g_token_'+user_email, user_info, ex=1 * 60 * 10)
+        return '/user/fill_details?g_token='+data['sub']+'&email='+user_email
```
Comment (raj454raj) [inline modules/utilities.py:89]: use URL helper

File: views/default/user.html (hunk 13 above)
Comment (raj454raj) [reply views/default/user.html:28]: Even if we want to go ahead with hardcoded image we will need separate buttons for sign in and sign up

File: controllers/default.py [hunk 18]
```diff
@@ -975,6 +975,37 @@ def user():
 
     return dict(form=auth())
 
+# ----------------------------------------------------------------------------
+
```
Comment (raj454raj) [inline controllers/default.py:979]:
```suggestion
```

File: controllers/user.py [hunk 19]
```diff
@@ -40,6 +40,66 @@ def login_token():
...
+    """
+    auth_token = request.vars.get('g_token')
+    email = request.vars.get('email')
+
+    #verify parameters
+    if not email or not auth_token:
+        raise HTTP(400, 'Invalid parameters')
+        return
+    #verify credentials
+    user_info = json.loads(current.REDIS_CLIENT.get("g_token_"+email))
```
Comment (raj454raj) [inline controllers/user.py:56]:
```suggestion
    user_info = json.loads(current.REDIS_CLIENT.get("g_token_" + email))
```

**Review by raj454raj — CHANGES_REQUESTED** [review]
@manav1403 still missed a few previous comments.

File: views/user/fill_details.html [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+{{block head}}
+  <title>{{=T("Create Profile")}}</title>
+  <meta name="description" content="Create profile on StopStalk">
+  <script>
+    $(document).ready(function() {
+      $('select').material_select();
+    });
+  </script>
+{{end}}
+<h1>{{=T("Create Profile")}}</h1>
```
Comment (raj454raj) [inline views/user/fill_details.html:11]:
```suggestion
<h1>{{=T("You're almost there!")}}</h1>
```

File: modules/utilities.py (hunk 5 above)
Comment (manav1403) [reply modules/utilities.py:59]: the API requests need to always return JSON response
writing this here will save us from appending the .json extension in the request urls

**Review by raj454raj — APPROVED** [review]
